pkg/plugin/plugin.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2018 Red Hat, Inc.
+// Copyright 2014 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Go version 1.10 or greater is required. Before that, switching namespaces in
+// long running processes in go did not work in a reliable way.
+// +build go1.10
+
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os/exec"
+	"runtime"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/vishvananda/netlink"
+)
+
+type netConf struct {
+	types.NetConf
+	BrName  string `json:"bridge"`
+	VlanTag *uint  `json:"vlan"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func logCall(command string, args *skel.CmdArgs) {
+	log.Printf("CNI %s was called for container ID: %s, network namespace %s, interface name %s, configuration: %s",
+		command, args.ContainerID, args.Netns, args.IfName, string(args.StdinData[:]))
+}
+
+func assertOvsAvailable() error {
+	// ovs-vsctl show will fail if OVS is not installed, running or user does
+	// not have rights to use it
+	if err := exec.Command("ovs-vsctl", "show").Run(); err != nil {
+		return fmt.Errorf("Open vSwitch is not available: %v", err)
+	}
+	return nil
+}
+
+func loadNetConf(bytes []byte) (*netConf, error) {
+	netconf := &netConf{}
+	if err := json.Unmarshal(bytes, netconf); err != nil {
+		return nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	if netconf.BrName == "" {
+		return nil, fmt.Errorf("\"bridge\" is a required argument")
+	}
```

Comment:
Also can we make a vlan validation?

Reply:
It is later validated by ovs-vsctl `ovs-vsctl: constraint violation: 5555 is not in the valid range 0 to 4095 (inclusive)`, I think we should validate only necessary minimum, but I'm open to discuss it.